net/gcoap: split out the coap processing part of _on_sock_evt

janosbrodbeck opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This PR splits out the processing part of _on_sock_evt into its own function.

At the moment the pdu processing is completely tied to events on UDP sockets. For potential upcoming DTLS support there will also be separate DTLS socket events. You could probably unpack the DTLS socket events to pass them to the UDP event handler, but I don't think this is a clean solution. The (big) processing part of the coap pdu and events on the socket handlers should be separated.

I have pulled this commit out of my current approaches to DTLS support in gcoap. Yes, the current parameters of these functions are also UDP exclusive, but I want to keep this a more or less simple change. Changes on the parameters would then come with the DTLS pull request. It still makes it easier when it is already separated.

It's just an proposal to split these parts this way, other suggestions are more than welcome. 


<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
The same procedures that test gcoap currently. It should hopefully not change anything functional.
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

janosbrodbeck commented:
Ping @chrysn since he is currently also working on gcoap.

chrysn commented:
already on it :-)

chrysn reviewed (commented):
It's a good refactoring step towards supporting more than just CoAP-over-UDP, and (after some battling the diff viewer) really just splitting things up. (The odd closed-brace at the end stems from the original `if (messagelayer_emptyresponse_type != NO_IMMEDIATE_REPLY) {` line being one level too far outside, logically, and being pulled into the coap-part of the function where it belongs.)

Clearing for CI; if there's a reason for the arg (see above) or it's removed, and the CI is happy, this can go ahead.

Tested on native and nrf52840dongle using gcoap example and `coap get ::1 5683 /.well-known/core`.

janosbrodbeck commented:
> It's a good refactoring step towards supporting more than just CoAP-over-UDP, and (after some battling the diff viewer) really just splitting things up. (The odd closed-brace at the end stems from the original `if (messagelayer_emptyresponse_type != NO_IMMEDIATE_REPLY) {` line being one level too far outside, logically, and being pulled into the coap-part of the function where it belongs.)
> 
> Clearing for CI; if there's a reason for the arg (see above) or it's removed, and the CI is happy, this can go ahead.
> 
> Tested on native and nrf52840dongle using gcoap example and `coap get ::1 5683 /.well-known/core`.

Removed the arg of `_process_coap_pdu`, was just a forgotten leftover from my DTLS changes. 

Since you mentioned refactoring: I have added another commit for which a separate PR is not worth it. Simply renaming `sock_udp_t _sock` to `sock_udo_t _sock_udp`. Since we have now a `_on_sock_udp_evt` function it may already make sense to declare the socket by socket type. But also fine with removing it again if unwanted.

janosbrodbeck commented:
Please don't merge this pull request! 

I want to reconsider one small part of it. I don't want to close it - just need a little time for it! Sorry for that :/

janosbrodbeck commented:
I've reverted the name change of `on_sock_evt` (renamed it to `on_sock_udp_evt` before). That was one step too far for me at the moment.

Can go on from my side

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -45,6 +45,8 @@
 /* Internal functions */
 static void *_event_loop(void *arg);
 static void _on_sock_evt(sock_udp_t *sock, sock_async_flags_t type, void *arg);
+static void _process_coap_pdu(sock_udp_t *sock, sock_udp_ep_t *remote,
+                            uint8_t *buf, size_t len);
```
```suggestion
static void _process_coap_pdu(sock_udp_t *sock, sock_udp_ep_t *remote,
                              uint8_t *buf, size_t len);
```

janosbrodbeck replied:
I'm confused. There needs to be a whitespace? Tab indents are not enough?

benpicco replied:
just so it aligns with the open brace, thought that looks tidier ¯\\_(ツ)_/¯

chrysn replied:
There's never any tabs in RIOT code, just spaces.

janosbrodbeck replied:
Better? Thats neat and similar to some other functions now.

janosbrodbeck replied:
(`_find_obs_memo_resource` could need one mir whitespace too now imho)

chrysn replied:
Yes, but that better goes into whenever that's actually touched next, or someone does a dedicated large-scale whitespace cleanup.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -119,24 +121,41 @@ static void *_event_loop(void *arg)
+}
+
+/* Processes and evaluates the coap pdu */
+static void _process_coap_pdu(sock_udp_t *sock, sock_udp_ep_t *remote,
+                            uint8_t *buf, size_t len)
```
```suggestion
static void _process_coap_pdu(sock_udp_t *sock, sock_udp_ep_t *remote,
                              uint8_t *buf, size_t len)
```

benpicco reviewed (approved):
looks good to me - just squash that whitespace fix in real quick :wink:

miri64 commented:
:rocket:
